### PR TITLE
Attempts to get Iron working...

### PR DESCRIPTION
**Pull request by thorncp**
Changes to get the worker successfully communicating with my local Hound. There's aren't final and the tests do not pass. I think they should be updated to stub the actual http request instead of expectations on Faraday.

https://trello.com/c/mfRjUISp

### CODE REVIEW

File: lib/scss_worker/worker.rb [hunk 1]
```diff
@@ -31,18 +31,25 @@ def run
...
+        client.token_auth(payload.fetch("token"))
+        client.adapter Faraday.default_adapter
       end
+
+      response = hound_client.put do |request|
+        request.url payload.fetch("hound_url")
+        request.body = hound_payload(violations)
+      end
+
+      p response.body
```
Comment (thorncp) [inline lib/scss_worker/worker.rb]: This isn't even a rails app!

Comment (srt32) [inline lib/scss_worker/worker.rb]: Oh, strange!

Comment (gylaz) [inline lib/scss_worker/worker.rb]: I think we still want to be notified of left-over debug statements.

**Comment by thorncp** [comment]
Update to get the specs passing. The tests only cover the happy path and don't exhaust any edge cases or failure states. I think it'll make sense to to split the worker up into several smaller objects with thorough testing, but this current state *might* be good enough to move forward with testing on staging. At the moment, Hound proper still has failing specs, but this implementation of the worker can successfully communicate with Hound. [Example](https://github.com/thornco/dumb-ruby/pull/33)

Summary from the commit:

> * Use webmock to make assertions on http requests that are made
> * Use faraday_middleware to send requests to Hound as JSON instead of
>   default of URL encoding
> * Group violations by line number to reduce payload size

File: lib/scss_worker/worker.rb [hunk 2]
```diff
@@ -31,23 +37,30 @@ def run
...
 
     def post_back_to_hound(violations)
-      Faraday.post do |request|
-        request.url = hound_url
-        request.token_auth(payload.fetch("token"))
-        request.body = hound_payload(violations).to_json
+      hound_client = Faraday.new do |client|
+        client.request :json
+        client.token_auth(payload.fetch("token"))
+        client.adapter :net_http
```
Comment (croaky) [inline lib/scss_worker/worker.rb]: `:sort`

Comment (croaky) [inline lib/scss_worker/worker.rb]: Any benefit to using some other adapter?

File: lib/scss_worker/worker.rb [hunk 3]
```diff
@@ -31,23 +37,30 @@ def run
...
         build_id: payload.fetch("build_id"),
-        violations: violations.to_json,
+        violations: violations,
         file: {
           name: payload.fetch("file").fetch("name"),
           content: payload.fetch("file").fetch("content"),
-          patch_body: payload.fetch("file").fetch("patch_body")
-        }
+          patch_body: payload.fetch("file").fetch("patch_body"),
+        },
```
Comment (croaky) [inline lib/scss_worker/worker.rb]: Sort keys in outer and inner hash?

File: spec/scss_worker/worker_spec.rb [hunk 4]
```diff
@@ -1,59 +1,90 @@
...
-          expect(Faraday).to have_received(:post)
-          expect(request).to have_received(:url=).with(request_stub[:hound_url])
-          expect(request).
-            to have_received(:token_auth).with(request_stub[:token])
-          expect(request).to have_received(:body=).with(response.to_json)
+          expect(request).to have_been_requested
         end
       end
+
+      # TODO: handle failures
```
Comment (croaky) [inline spec/scss_worker/worker_spec.rb]: What are the expected failure cases?

* Poorly formatted config (bad YAML/JSON)
* Keys that don't exist as options in SCSS Lint
* 4xx response - client error, our fault
* 5xx response - server error, Hound's fault or general networking error

**Comment by croaky** [comment]
I think the happy path is good for this PR to be merged and deployed, then we enumerate the failure cases and get Trello cards in the backlog for each one.

File: lib/scss_worker/worker.rb (hunk 2 above)
Comment (thorncp) [inline lib/scss_worker/worker.rb]: This is setting up a Rack middleware stack, so the order is somewhat important. For whatever reason, `client.request :json` has to be first, or the specs will fail, but I sorted the other two in 556dfdfc130d6cddb14a5d6a59f14647f4016828. So far Faraday has been a headache more than anything else :unamused:

Comment (thorncp) [inline lib/scss_worker/worker.rb]: I don't know if there's a benefit for other adapters or not, but it is something we can look into.

File: lib/scss_worker/worker.rb (hunk 3 above)
Comment (thorncp) [inline lib/scss_worker/worker.rb]: Done in 556dfdfc130d6cddb14a5d6a59f14647f4016828

File: spec/scss_worker/worker_spec.rb (hunk 4 above)
Comment (thorncp) [inline spec/scss_worker/worker_spec.rb]: Looks like a good list to me!

File: lib/scss_worker/worker.rb [hunk 5]
```diff
@@ -31,23 +37,30 @@ def run
     private
 
     def post_back_to_hound(violations)
-      Faraday.post do |request|
-        request.url = hound_url
-        request.token_auth(payload.fetch("token"))
-        request.body = hound_payload(violations).to_json
+      hound_client = Faraday.new do |client|
```
Comment (srt32) [inline lib/scss_worker/worker.rb]: Is it worth pulling `hound_client` into a private method to shorten up this method?

File: spec/scss_worker/worker_spec.rb [hunk 6]
```diff
@@ -1,59 +1,90 @@
...
+    def stub_callback_request(violations)
+      stub_request(:put, build_worker_callback_url).with(
+        body: {
+          build_id: build_id,
+          build_worker_id: build_worker_id,
+          file: stub_file,
+          violations: violations,
+        },
+        headers: {
+          Authorization: 'Token token="%s"' % hound_token,
```
Comment (srt32) [inline spec/scss_worker/worker_spec.rb]: Out of curiosity, why not use `#{}` style string interpolation?

**Comment by srt32** [comment]
Couple small questions. LGTM!

File: spec/scss_worker/worker_spec.rb (hunk 6 above)
Comment (thorncp) [inline spec/scss_worker/worker_spec.rb]: Because the header value requires the double quotes. Alternatives would be

* `"Token token=\"#{hound_token}\""`
* `%{Token token="#{hound_token}"}`

I suppose that second one doesn't look too bad.

Comment (srt32) [inline spec/scss_worker/worker_spec.rb]: Ah, yes. :ok_hand:

Comment (gylaz) [inline spec/scss_worker/worker_spec.rb]: Yep, we even have a guideline to prefer `%{}` in such cases: https://github.com/thoughtbot/guides/tree/master/style/ruby

Comment (thorncp) [inline spec/scss_worker/worker_spec.rb]: Updated in b71c141139a2c18ef49f7e51764f8dc38be13e04

**Comment by gylaz** [comment]
Is this good to merge, or are we waiting on something else to land?
